lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -3900,14 +3919,17 @@ ADContext::declareExternalAssociatedFunction(
       break;
   }
   auto assocGenSig = getAssociatedFunctionGenericSignature(attr, original);
+  auto *assocGenEnv = assocGenSig
```

Comment:
Are you going to synthesize a body for the function? Then it doesn't need a generic environment. We should enforce this with the SIL verifier.

Reply:
No, actually, no body is synthesized for `assocFn` here.

If I don't create `assocFn` with its generic environment, then I can't access its substitution map in `reapplyFunctionConversion` (where our other conversation lives).

Reply:
Which substitution map are you unable to access if there's no generic environment? I can't find anything that looks suspicious.

Reply:
The code here (`ADContext::declareExternalAssociatedFunction`) currently constructs `assocFn` with a generic environment and returns `assocFn`.

`assocFn` is then:
* Stored in a `DifferentiationTask`,
* Retrieved [here](https://github.com/apple/swift/pull/22240/files#diff-4ede198777b59748c377cdd17c444e83R1761) via `task->getAssociatedFunction(kind)`...
* So that its substitution map can be passed to `reapplyFunctionConversion` as argument `substMap`, so it can be combined with `pai->getSubstitutionMap()`.

---

If I construct `assocFn` without a generic environment, then the `substMap` passed to `reapplyFunctionConversion` is always empty and I'm not able to access derivative functions' substitution maps.

Reply:
Changed to suggested substitution map logic in https://github.com/apple/swift/pull/22240/commits/8b68594cded4de1ede729c7b4abe4575ddea81e0.